src/ProductConstructionService/ProductConstructionService.Api/Queue/JobsProcessorScopeManager.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace ProductConstructionService.Api.Queue;
+
+public enum JobsProcessorState
+{
+    // The JobsProcessor will keep taking and processing new jobs
+    Working,
+    // The JobsProcessor isn't doing anything
+    Stopped,
+    // The JobsProcessor will finish it's current job and stop
+    Stopping
+}
+
+public class JobsProcessorScopeManager
+{
+    public JobsProcessorScopeManager()
+    {
+        State = JobsProcessorState.Stopped;
+        _autoResetEvent = new(false);
+    }
+
+    public JobsProcessorState State { get; private set; }
+    private readonly AutoResetEvent _autoResetEvent;
+
+    public void Start()
+    {
+        State = JobsProcessorState.Working;
+        _autoResetEvent.Set();
+    }
+
+    /// <summary>
+    /// Creates a new scope for the currently executing Job, when the the JobsProcessor is in the `Working` state.
+    /// </summary>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    public IDisposable BeginJobScopeWhenReady(CancellationToken cancellationToken)
+    {
+        _autoResetEvent.WaitOne();
```

Comment:
I think I see a problem.

What if:
- There are no jobs
- We're waiting for one (we're working)
- We want to deploy, we call `Stop()`
- Now, we will stay forever `Stopping` until we get a job?

What I mean is that this `WaitOne()` will keep hanging until it gets the next job and only then we will be able to stop fully.
So the deployment will never finish.

Is this correct?

Reply:
No, when there are no jobs, we try getting a job, fail, sleep for 60 seconds, and then exit the using scope, so `Dispose` will be called, it will see that we're `Stopping`, and it will flip to `Stop` without settings the `autoResetEvent`, blocking us in the next "circle"

Reply:
Ah, great!